From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: gellert <gellert.hegyi@around.co>
Date: Fri, 13 Aug 2021 11:42:52 +0200
Subject: linux add window will-resize did-resize events

This patch adds similar window did-resize and will-resize events to linux as there is for macOS and Windows.

diff --git a/ui/aura/window_tree_host.cc b/ui/aura/window_tree_host.cc
index 27237fb845ab84de04f84eddcd5b79f1c784489d..17ac1d4b1177ff3bc79023cc16450b6ac76f76e0 100644
--- a/ui/aura/window_tree_host.cc
+++ b/ui/aura/window_tree_host.cc
@@ -550,6 +550,24 @@ void WindowTreeHost::OnHostMovedInPixels(
     observer.OnHostMovedInPixels(this, new_location_in_pixels);
 }
 
+void WindowTreeHost::OnHostWillResize(const gfx::Size& new_size_in_pixels) {
+  if (!compositor_)
+    return;
+  display::Display display =
+      display::Screen::GetScreen()->GetDisplayNearestWindow(window());
+  device_scale_factor_ = display.device_scale_factor();
+  UpdateRootWindowSizeInPixels();
+  UpdateCompositorScaleAndSize(new_size_in_pixels);
+
+  for (WindowTreeHostObserver& observer : observers_)
+    observer.OnHostWillResize(this);
+}
+
+void WindowTreeHost::OnHostDidResize() {
+  for (WindowTreeHostObserver& observer : observers_)
+    observer.OnHostDidResize(this);
+}
+
 void WindowTreeHost::OnHostResizedInPixels(
     const gfx::Size& new_size_in_pixels) {
   // The compositor is deleted from WM_DESTROY, but we don't delete things until
diff --git a/ui/aura/window_tree_host.h b/ui/aura/window_tree_host.h
index 6e449e363660b6697ec96ca32d4c2d8a5579e5b3..a898289086676b3da13edaf72c85f54c3b266910 100644
--- a/ui/aura/window_tree_host.h
+++ b/ui/aura/window_tree_host.h
@@ -302,6 +302,10 @@ class AURA_EXPORT WindowTreeHost : public ui::internal::InputMethodDelegate,
 
   void OnHostMovedInPixels(const gfx::Point& new_location_in_pixels);
   void OnHostResizedInPixels(const gfx::Size& new_size_in_pixels);
+
+  void OnHostWillResize(const gfx::Size& new_size_in_pixels);
+  void OnHostDidResize();
+
   void OnHostWorkspaceChanged();
   void OnHostDisplayChanged();
   void OnHostCloseRequested();
diff --git a/ui/aura/window_tree_host_observer.h b/ui/aura/window_tree_host_observer.h
index 995ada6aa5d53f76f1672cb95040299722da8fd1..ad79d458028104c4a285c3abe88ca740e2e8be88 100644
--- a/ui/aura/window_tree_host_observer.h
+++ b/ui/aura/window_tree_host_observer.h
@@ -22,6 +22,9 @@ class AURA_EXPORT WindowTreeHostObserver {
   // Called when the host's client size has changed.
   virtual void OnHostResized(WindowTreeHost* host) {}
 
+  virtual void OnHostWillResize(WindowTreeHost* host) {}
+  virtual void OnHostDidResize(WindowTreeHost* host) {}
+
   // Called when the host is moved on screen.
   virtual void OnHostMovedInPixels(WindowTreeHost* host,
                                    const gfx::Point& new_origin_in_pixels) {}
diff --git a/ui/aura/window_tree_host_platform.cc b/ui/aura/window_tree_host_platform.cc
index e793e0812cfaa765417c8da8afd13cae4ced0b64..4bea4ab9156f13cdfd3081c783f8f0a7a31778b2 100644
--- a/ui/aura/window_tree_host_platform.cc
+++ b/ui/aura/window_tree_host_platform.cc
@@ -208,6 +208,16 @@ void WindowTreeHostPlatform::LockMouse(Window* window) {
   WindowTreeHost::LockMouse(window);
 }
 
+#if defined(OS_LINUX)
+void WindowTreeHostPlatform::OnWillResize(const gfx::Rect& new_bounds) {
+  OnHostWillResize(bounds_in_pixels_.size());
+}
+
+void WindowTreeHostPlatform::OnDidResize() {
+  OnHostDidResize();
+}
+#endif
+
 void WindowTreeHostPlatform::OnBoundsChanged(const BoundsChange& change) {
   // It's possible this function may be called recursively. Only notify
   // observers on initial entry. This way observers can safely assume that
diff --git a/ui/aura/window_tree_host_platform.h b/ui/aura/window_tree_host_platform.h
index 40ebc2109f5f7c822979d3861f34e9da7e283645..6b2a4c52bd0cbd699a8e69f11ff95e7fc429fe14 100644
--- a/ui/aura/window_tree_host_platform.h
+++ b/ui/aura/window_tree_host_platform.h
@@ -72,6 +72,12 @@ class AURA_EXPORT WindowTreeHostPlatform : public WindowTreeHost,
 
   // ui::PlatformWindowDelegate:
   void OnBoundsChanged(const BoundsChange& change) override;
+
+#if defined(OS_LINUX)
+  void OnWillResize(const gfx::Rect& new_bounds) override;
+  void OnDidResize() override;
+#endif
+
   void OnDamageRect(const gfx::Rect& damaged_region) override;
   void DispatchEvent(ui::Event* event) override;
   void OnCloseRequest() override;
diff --git a/ui/platform_window/platform_window_delegate.h b/ui/platform_window/platform_window_delegate.h
index 73b1ef6cec862e7a86e0986be585afbf7c0846f2..d0c811f931b4674d468d41fc4bc3922906eddf14 100644
--- a/ui/platform_window/platform_window_delegate.h
+++ b/ui/platform_window/platform_window_delegate.h
@@ -75,6 +75,11 @@ class COMPONENT_EXPORT(PLATFORM_WINDOW) PlatformWindowDelegate {
 
   virtual void OnBoundsChanged(const BoundsChange& change) = 0;
 
+#if defined(OS_LINUX)
+  virtual void OnWillResize(const gfx::Rect& new_bounds) = 0;
+  virtual void OnDidResize() = 0;
+#endif
+
   // Note that |damaged_region| is in the platform-window's coordinates, in
   // physical pixels.
   virtual void OnDamageRect(const gfx::Rect& damaged_region) = 0;
diff --git a/ui/platform_window/x11/x11_window.cc b/ui/platform_window/x11/x11_window.cc
index 3ef183213bc604656d396abf5af535818f8c2833..815299e71725bae5784c69da1e7cad3dd229b92d 100644
--- a/ui/platform_window/x11/x11_window.cc
+++ b/ui/platform_window/x11/x11_window.cc
@@ -1373,6 +1373,14 @@ void X11Window::OnXWindowBoundsChanged(const gfx::Rect& bounds) {
   platform_window_delegate_->OnBoundsChanged(bounds);
 }
 
+void X11Window::OnXWindowWillResize(const gfx::Rect& bounds) {
+  platform_window_delegate_->OnWillResize(bounds);
+}
+
+void X11Window::OnXWindowDidResize() {
+  platform_window_delegate_->OnDidResize();
+}
+
 void X11Window::OnXWindowCloseRequested() {
   platform_window_delegate_->OnCloseRequest();
 }
@@ -2077,6 +2085,19 @@ void X11Window::HandleEvent(const x11::Event& xev) {
                                     expose->height);
     OnXWindowDamageEvent(damage_rect_in_pixels);
   } else if (auto* focus = xev.As<x11::FocusEvent>()) {
+    if (focus->mode == x11::NotifyMode::Grab) {
+      is_mouse_button_down_ = true;
+    }
+
+    if (focus->mode == x11::NotifyMode::Ungrab) {
+      is_mouse_button_down_ = false;
+
+      if (is_resizing_) {
+        is_resizing_ = false;
+        OnXWindowDidResize();
+      }
+    }
+
     OnFocusEvent(focus->opcode == x11::FocusEvent::In, focus->mode,
                  focus->detail);
   } else if (auto* configure = xev.As<x11::ConfigureNotifyEvent>()) {
@@ -2323,6 +2344,13 @@ void X11Window::DelayedResize(const gfx::Rect& bounds_in_pixels) {
                    ++current_counter_value_);
   }
 
+  if (is_mouse_button_down_) {
+    if (!is_resizing_) {
+      OnXWindowWillResize(bounds_in_pixels);
+    }
+    is_resizing_ = true;
+  }
+
   CancelResize();
   NotifyBoundsChanged(bounds_in_pixels);
 
diff --git a/ui/platform_window/x11/x11_window.h b/ui/platform_window/x11/x11_window.h
index 67482087785d995765e056d136d88b2fe7495c42..665ee1c2385c558815d2b3022e4f3f7643b136c9 100644
--- a/ui/platform_window/x11/x11_window.h
+++ b/ui/platform_window/x11/x11_window.h
@@ -147,6 +147,10 @@ class X11_WINDOW_EXPORT X11Window : public PlatformWindow,
   void OnXWindowStateChanged();
   void OnXWindowDamageEvent(const gfx::Rect& damage_rect);
   void OnXWindowBoundsChanged(const gfx::Rect& size);
+
+  void OnXWindowWillResize(const gfx::Rect& size);
+  void OnXWindowDidResize();
+
   void OnXWindowCloseRequested();
   void OnXWindowIsActiveChanged(bool active);
   void OnXWindowWorkspaceChanged();
@@ -497,6 +501,9 @@ class X11_WINDOW_EXPORT X11Window : public PlatformWindow,
 
   base::CancelableOnceCallback<void(x11::Cursor)> on_cursor_loaded_;
 
+  bool is_resizing_ = false;
+  bool is_mouse_button_down_ = false;
+
   base::WeakPtrFactory<X11Window> weak_ptr_factory_{this};
 };
 
diff --git a/ui/views/widget/desktop_aura/desktop_native_widget_aura.cc b/ui/views/widget/desktop_aura/desktop_native_widget_aura.cc
index 43cbc4940c2a5b698c5c8b7453cc916a362c5c42..a1091b9fb25a8e01810f9ba1401ca2186f95d415 100644
--- a/ui/views/widget/desktop_aura/desktop_native_widget_aura.cc
+++ b/ui/views/widget/desktop_aura/desktop_native_widget_aura.cc
@@ -1330,6 +1330,21 @@ void DesktopNativeWidgetAura::OnHostCloseRequested(aura::WindowTreeHost* host) {
   GetWidget()->Close();
 }
 
+#if defined(OS_LINUX)
+void DesktopNativeWidgetAura::OnHostWillResize(aura::WindowTreeHost* host) {
+  if (desktop_window_tree_host_->IsAnimatingClosed())
+    return;
+
+  gfx::Rect new_bounds = gfx::Rect(host->window()->bounds().size());
+  content_window_->SetBounds(new_bounds);
+  native_widget_delegate_->OnNativeWidgetWillResize(new_bounds.size());
+}
+
+void DesktopNativeWidgetAura::OnHostDidResize(aura::WindowTreeHost* host) {
+  native_widget_delegate_->OnNativeWidgetDidResize();
+}
+#endif
+
 void DesktopNativeWidgetAura::OnHostResized(aura::WindowTreeHost* host) {
   // Don't update the bounds of the child layers when animating closed. If we
   // did it would force a paint, which we don't want. We don't want the paint
diff --git a/ui/views/widget/desktop_aura/desktop_native_widget_aura.h b/ui/views/widget/desktop_aura/desktop_native_widget_aura.h
index a0cf6b85cc0c5ec0e307c8a077b830f51b1f043b..b6dd1e4b6cc618dde25c7e023dcc573a5db09164 100644
--- a/ui/views/widget/desktop_aura/desktop_native_widget_aura.h
+++ b/ui/views/widget/desktop_aura/desktop_native_widget_aura.h
@@ -263,6 +263,12 @@ class VIEWS_EXPORT DesktopNativeWidgetAura
   // aura::WindowTreeHostObserver:
   void OnHostCloseRequested(aura::WindowTreeHost* host) override;
   void OnHostResized(aura::WindowTreeHost* host) override;
+
+#if defined(OS_LINUX)
+  void OnHostWillResize(aura::WindowTreeHost* host) override;
+  void OnHostDidResize(aura::WindowTreeHost* host) override;
+#endif
+
   void OnHostWorkspaceChanged(aura::WindowTreeHost* host) override;
   void OnHostMovedInPixels(aura::WindowTreeHost* host,
                            const gfx::Point& new_origin_in_pixels) override;
diff --git a/ui/views/widget/native_widget_delegate.h b/ui/views/widget/native_widget_delegate.h
index fbfefa1e8e5bebdaab2a3073c9c5618113293755..45e26ea5ff143b57afed40b28be0d1cb3573b6bf 100644
--- a/ui/views/widget/native_widget_delegate.h
+++ b/ui/views/widget/native_widget_delegate.h
@@ -87,6 +87,11 @@ class VIEWS_EXPORT NativeWidgetDelegate {
   // e.g. maximize.
   virtual void OnNativeWidgetSizeChanged(const gfx::Size& new_size) = 0;
 
+#if defined(OS_LINUX)
+  virtual void OnNativeWidgetWillResize(const gfx::Size& new_size) = 0;
+  virtual void OnNativeWidgetDidResize() = 0;
+#endif
+
   // Called when NativeWidget changed workspaces or its visible on all
   // workspaces state changes.
   virtual void OnNativeWidgetWorkspaceChanged() = 0;
diff --git a/ui/views/widget/widget.cc b/ui/views/widget/widget.cc
index bbbf0478947b941c30b4086cdca3871ee51e5e04..79f6c6c2c656cb22bf96522f4b67e6ab79c821d4 100644
--- a/ui/views/widget/widget.cc
+++ b/ui/views/widget/widget.cc
@@ -1431,6 +1431,25 @@ void Widget::OnNativeWidgetMove() {
     observer.OnWidgetBoundsChanged(this, GetWindowBoundsInScreen());
 }
 
+#if defined(OS_LINUX)
+void Widget::OnNativeWidgetWillResize(const gfx::Size& new_size) {
+  View* root = GetRootView();
+  if (root)
+    root->SetSize(new_size);
+
+  NotifyCaretBoundsChanged(GetInputMethod());
+  SaveWindowPlacementIfInitialized();
+
+  for (WidgetObserver& observer : observers_)
+    observer.OnWidgetWillResize(this, GetWindowBoundsInScreen());
+}
+
+void Widget::OnNativeWidgetDidResize() {
+  for (WidgetObserver& observer : observers_)
+    observer.OnWidgetDidResize(this);
+}
+#endif
+
 void Widget::OnNativeWidgetSizeChanged(const gfx::Size& new_size) {
   TRACE_EVENT0("ui", "Widget::OnNativeWidgetSizeChanged");
 
diff --git a/ui/views/widget/widget.h b/ui/views/widget/widget.h
index 31d6c01a839baf217e63bbbc592920cc88f61e5b..47302aa8190856464d8f3fb57938d2c04f3f5623 100644
--- a/ui/views/widget/widget.h
+++ b/ui/views/widget/widget.h
@@ -1006,6 +1006,12 @@ class VIEWS_EXPORT Widget : public internal::NativeWidgetDelegate,
   gfx::Size GetMaximumSize() const override;
   void OnNativeWidgetMove() override;
   void OnNativeWidgetSizeChanged(const gfx::Size& new_size) override;
+
+#if defined(OS_LINUX)
+  void OnNativeWidgetWillResize(const gfx::Size& new_size) override;
+  void OnNativeWidgetDidResize() override;
+#endif
+
   void OnNativeWidgetWorkspaceChanged() override;
   void OnNativeWidgetWindowShowStateChanged() override;
   void OnNativeWidgetBeginUserBoundsChange() override;
diff --git a/ui/views/widget/widget_observer.h b/ui/views/widget/widget_observer.h
index 3a8cf30252ee732ff9364383855753ef359d3c0b..448337145a99d5a5da7adb3c6c7d528df07ca5cf 100644
--- a/ui/views/widget/widget_observer.h
+++ b/ui/views/widget/widget_observer.h
@@ -53,6 +53,9 @@ class VIEWS_EXPORT WidgetObserver : public base::CheckedObserver {
   virtual void OnWidgetBoundsChanged(Widget* widget,
                                      const gfx::Rect& new_bounds) {}
 
+  virtual void OnWidgetWillResize(Widget* widget, const gfx::Rect& new_bounds) {}
+  virtual void OnWidgetDidResize(Widget* widget) {}
+
   virtual void OnWidgetThemeChanged(Widget* widget) {}
 
  protected:
